dadi/lib/auth/index.js

```diff
@@ -1,184 +1,129 @@
-    if (req.url === tokenRoute || !mustAuthenticate(server.components, req)) return next()
+  error.statusCode = 401
 
-    // require an authorization header for every request
-    if (!(req.headers && req.headers.authorization)) return fail()
+  res.setHeader(
+    'WWW-Authenticate',
+    'Bearer, error="invalid_credentials", error_description="Invalid credentials supplied"'
+  )
 
-    // Strip token value out of request headers
-    const parts = req.headers.authorization.split(' ')
+  next(error)
+}
 
-    let token
+/**
+ * Takes the client ID and secret present in the request body and
+ * attempts to generate a bearer token with them. If successful,
+ * a signed JWT will be sent in the response, as well as some additional
+ * data about the client. If not, a 401 error is delivered.
+ *
+ * @param  {http.IncomingMessage}   req
+ * @param  {http.ServerResponse}    res
+ * @param  {Function}               next
+ */
+AuthMiddleware.prototype.generateToken = function (req, res, next) {
+  let {clientId, secret} = req.body
+
+  res.setHeader('Content-Type', 'application/json')
+  res.setHeader('Cache-Control', 'no-store')
+  res.setHeader('Pragma', 'no-cache')
+
+  if (
+    typeof clientId !== 'string' ||
+    typeof secret !== 'string'
+  ) {
+    return this.handleInvalidCredentials(req, res, next)
+  }
 
-    // Headers should be `Authorization: Bearer <%=tokenvalue%>`
-    if (parts.length === 2 && /^Bearer$/i.test(parts[0])) {
-      token = parts[1]
+  clientModel.get(clientId, secret).then(({results}) => {
+    if (results.length === 0) {
+      return this.handleInvalidCredentials(req, res, next)
     }
 
-    if (!token) return fail('NoToken')
+    let client = clientModel.sanitise(results[0])
+    let payload = {
+      clientId,
+      accessType: client.accessType
+    }
 
-    tokens.validate(token, function (err, client) {
+    jwt.sign(payload, config.get('auth.tokenKey'), {
+      expiresIn: config.get('auth.tokenTtl')
+    }, (err, token) => {
       if (err) {
-        if (err.message === 'DB_DISCONNECTED') {
-          const error = formatError.createApiError('0004')
-
-          error.statusCode = 503
-
-          return next(error)
-        }
-
-        return next(err)
+        return this.handleInvalidCredentials(req, res, next)
       }
 
-      // If token is good continue, else `fail()`
-      if (client) {
-        if (!isAuthorized(server.components, req, client)) {
-          const error = new Error('ClientId not authorized to access requested collection.')
-          error.statusCode = 401
-
-          res.setHeader('WWW-Authenticate', 'Bearer realm="' + url.parse(req.url, true).pathname + '"')
-
-          return next(error)
-        } else {
-          // Token is valid. Attach client to request.
-          req.client = client
-
-          return next()
+      let response = {
+        accessToken: token,
+        tokenType: 'Bearer',
+        expiresIn: config.get('auth.tokenTtl'),
```

Comment:
What do you think about implementing the same human-readable expiry here as we do for media signed URLs?

Reply:
We could introduce support for that format in the config, so that you could type in `'12h'` instead of `43200`. However, I wouldn't use that in the `expiresIn` property of the token endpoint response, because this property will most likely be used programmatically by consumer applications to determine how long a bearer token is valid for (e.g. [Passport](https://github.com/api/passport) or [API Wrapper](https://github.com/dadi/api-wrapper)). Changing that property to be human-readable would break those applications.

The only solution would be to add another property, which would be a human-readable version of `expiresIn`, but to be completely honest I don't think it's worth the hassle.